Makefile

```diff
@@ -78,91 +113,572 @@ verify:
+#    part of this test run.  Defaults to false.
+#  ARTIFACTS: For REMOTE=true only.  Local directory to scp test artifacts into
+#    from the remote hosts.  Defaults to ""/tmp/_artifacts".
+#  REPORT: For REMOTE=false only.  Local directory to write juntil xml results
+#    to.  Defaults to "/tmp/".
+#  CLEANUP: For REMOTE=true only.  If false, do not stop processes or delete
+#    test files on remote hosts.  Defaults to true.
+#  IMAGE_PROJECT: For REMOTE=true only.  Project containing images provided to
+#  IMAGES.  Defaults to "kubernetes-node-e2e-images".
+#  INSTANCE_PREFIX: For REMOTE=true only.  Instances created from images will
+#    have the name "${INSTANCE_PREFIX}-${IMAGE_NAME}".  Defaults to "test".
+#
+# Example:
+#   make test-e2e-node FOCUS=kubelet SKIP=container
+#   make test-e2e-node REMOTE=true DELETE_INSTANCES=true
 # Build and run tests.
-test_e2e_node: ginkgo
-	hack/e2e-node-test.sh
-.PHONY: test_e2e_node
+.PHONY: test-e2e-node
+test-e2e-node: ginkgo generated_files
+	hack/make-rules/test-e2e-node.sh
+
+# Build and run cmdline tests.
+#
+# Example:
+#   make test-cmd
+test-cmd:
+	@hack/make-rules/test-cmd.sh
+.PHONY: test-cmd
 
 # Remove all build artifacts.
 #
 # Example:
 #   make clean
-clean:
+.PHONY: clean
+clean: clean_generated clean_meta
 	build/make-clean.sh
 	rm -rf $(OUT_DIR)
 	rm -rf Godeps/_workspace # Just until we are sure it is gone
-.PHONY: clean
+
+# Remove make-related metadata files.
+#
+# Example:
+#   make clean_meta
+.PHONY: clean_meta
+clean_meta:
+	rm -rf $(META_DIR)
+
+# Remove all auto-generated artifacts.
+#
+# Example:
+#   make clean_generated
+.PHONY: clean_generated
+clean_generated:
+	find . -type f -name $(GENERATED_FILE_PREFIX)\* | xargs rm -f
 
 # Run 'go vet'.
 #
 # Args:
 #   WHAT: Directory names to vet.  All *.go files under these
 #     directories will be vetted.  If not specified, "everything" will be
 #     vetted.
-#   TESTS: Same as WHAT.
-#   GOFLAGS: Extra flags to pass to 'go' when building.
-#   GOLDFLAGS: Extra linking flags to pass to 'go' when building.
 #
 # Example:
 #   make vet
 #   make vet WHAT=pkg/kubelet
-vet:
-	hack/verify-govet.sh $(WHAT) $(TESTS)
 .PHONY: vet
+vet:
+	hack/make-rules/vet.sh $(WHAT)
 
 # Build a release
 #
 # Example:
 #   make release
-release:
-	build/release.sh
 .PHONY: release
+release: generated_files
+	build/release.sh
 
 # Build a release, but skip tests
 #
 # Example:
 #   make release-skip-tests
-release-skip-tests quick-release:
-	KUBE_RELEASE_RUN_TESTS=n KUBE_FASTBUILD=true build/release.sh
 .PHONY: release-skip-tests quick-release
+release-skip-tests quick-release: generated_files
+	KUBE_RELEASE_RUN_TESTS=n KUBE_FASTBUILD=true build/release.sh
+
+# Cross-compile for all platforms
+#
+# Example:
+#   make cross
+.PHONY: cross
+cross:
+	hack/make-rules/cross.sh
+
+# Add rules for all directories in cmd/
+#
+# Example:
+#   make kubectl kube-proxy
+.PHONY: $(notdir $(abspath $(wildcard cmd/*/)))
```

Comment:
This was sort of gravy.  I could drop this, if needed.  Or extend to other bins.